### PR TITLE
resource_manager: impl limiter for token controlloer

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?

Issue Number: ref #5851 

### What is changed and how does it work?
Based on the rate limiter implemented by Go(ref to https://cs.opensource.google/go/x/time/+/refs/tags/v0.2.0:rate/rate.go), some modifications have been made.
The first is the introduction of token notification, which is sent to the channel when the threshold is below.
Second, only the `Reserve` method is retained, and `WaitReservations` takes example by `Wait` method  to handle multiple reservations. I also modified the reservation cancel function, once cancel, restore all the tokens.

```commit-message
impl limiter for token controlloer
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

**Review by nolouch — COMMENTED** [review]
plz add the changes item in the pull request description.

File: pkg/mcs/resource_manager/client/limiter.go [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+	r.lim.mu.Lock()
+	defer r.lim.mu.Unlock()
+
+	if r.lim.limit == Inf || r.tokens == 0 || r.timeToAct.Before(now) {
+		return
+	}
+	// advance time to now
+	now, _, tokens := r.lim.advance(now)
+	// calculate new number of tokens
+	tokens += float64(r.tokens)
```
Comment (nolouch) [inline pkg/mcs/resource_manager/client/limiter.go:153]: need address the lint check.

**Review by nolouch — COMMENTED** [review]
rest LGTM

File: pkg/mcs/resource_manager/client/limiter.go [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+			ok:        true,
+			lim:       lim,
+			tokens:    n,
+			timeToAct: now,
+		}
+	} else if n > maxRequestTokens {
+		return Reservation{
+			ok:        false,
+			lim:       lim,
+			tokens:    lim.tokens,
```
Comment (nolouch) [inline pkg/mcs/resource_manager/client/limiter.go:287]: ok is false, is the tokens field useless?

File: pkg/mcs/resource_manager/client/limiter.go [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+	defer lim.mu.Unlock()
+
+	if lim.limit == Inf {
+		return Reservation{
+			ok:        true,
+			lim:       lim,
+			tokens:    n,
+			timeToAct: now,
+		}
+	} else if n > maxRequestTokens {
```
Comment (nolouch) [inline pkg/mcs/resource_manager/client/limiter.go:283]: I think `maxRequestTokens` should be configuable , or no need maxRequestTokens  in client side.

File: pkg/mcs/resource_manager/client/limiter.go [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+	waitLimit := InfDuration
+	if deadline, ok := ctx.Deadline(); ok {
+		waitLimit = deadline.Sub(now)
+	}
+	r := lim.reserveN(now, n, waitLimit)
+	return &r
+}
+
+// SetupNotificationAt enables the notification at the given threshold.
+func (lim *Limiter) SetupNotificationAt(now time.Time, threshold float64) {
```
Comment (JmPotato) [inline pkg/mcs/resource_manager/client/limiter.go:193]: Using the word `At` is a little bit confusing for the caller. They may be confused about which parameter this `At` refers to.

```suggestion
func (lim *Limiter) SetupNotificationThreshold(now time.Time, threshold float64) {
```

File: pkg/mcs/resource_manager/client/limiter.go [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+// or its associated context.Context is canceled.
+type Limiter struct {
+	mu     sync.Mutex
+	limit  Limit
+	tokens float64
+	// last is the last time the limiter's tokens field was updated
+	last                time.Time
+	notifyThreshold     float64
+	lowTokensNotifyChan chan struct{}
+	isLowProcess        bool
```
Comment (JmPotato) [inline pkg/mcs/resource_manager/client/limiter.go:73]: What does this field mean? Adding a comment will be helpful.

**Review by nolouch — APPROVED** [review]
lgtm

**Review by JmPotato — APPROVED** [review]

**Pull request merged**